From ed7e79a42a23661a4551adcb5c3dae3164a8fca2 Mon Sep 17 00:00:00 2001
From: Doug Ledford <dledford@redhat.com>
Date: Wed, 24 Sep 2014 17:27:57 -0400
Subject: [infiniband] Return error for unsupported QP creation flags

Message-id: <6754ac83830c28837ddc97b3a157f3949c0a703a.1411578494.git.dledford@redhat.com>
Patchwork-id: 96675
O-Subject: [Patch RHEL7 04/26] IB: Return error for unsupported QP creation flags
Bugzilla: 1107621
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Or Gerlitz <ogerlitz@mellanox.com>

Fix the usnic and thw qib drivers to err when QP creation flags that
they don't understand are provided.

Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 60093dc0c8b6407bc7494cbcb3e84322cc6782c8)
Signed-off-by: Doug Ledford <dledford@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/qib/qib_qp.c b/drivers/infiniband/hw/qib/qib_qp.c
index 0cad0c4..7fcc150 100644
--- a/drivers/infiniband/hw/qib/qib_qp.c
+++ b/drivers/infiniband/hw/qib/qib_qp.c
@@ -985,7 +985,8 @@ struct ib_qp *qib_create_qp(struct ib_pd *ibpd,
  struct ib_qp *ret;
 
  if (init_attr->cap.max_send_sge > ib_qib_max_sges ||
-     init_attr->cap.max_send_wr > ib_qib_max_qp_wrs) {
+     init_attr->cap.max_send_wr > ib_qib_max_qp_wrs ||
+     init_attr->create_flags) {
   ret = ERR_PTR(-EINVAL);
   goto bail;
  }
diff --git a/drivers/infiniband/hw/usnic/usnic_ib_verbs.c b/drivers/infiniband/hw/usnic/usnic_ib_verbs.c
index d48d2c0..53bd6a2 100644
--- a/drivers/infiniband/hw/usnic/usnic_ib_verbs.c
+++ b/drivers/infiniband/hw/usnic/usnic_ib_verbs.c
@@ -466,6 +466,9 @@ struct ib_qp *usnic_ib_create_qp(struct ib_pd *pd,
  ucontext = to_uucontext(pd->uobject->context);
  us_ibdev = to_usdev(pd->device);
 
+ if (init_attr->create_flags)
+  return ERR_PTR(-EINVAL);
+
  err = ib_copy_from_udata(&cmd, udata, sizeof(cmd));
  if (err) {
   usnic_err("%s: cannot copy udata for create_qp\n",
-- 
1.7.1